### PR TITLE
Channel aware Product Primary Collections

### LINKED ISSUE #456: Primary collection plugin is not channel aware
For the ColourGraphics project, we have assigned the same product to multiple channels. A product can however only have 1 primary collection. In this setup a product should be able to ahve 2 primary collections: 1 for CG channel and 1 primary collection for the Fabrix channel.

How can we do this? Any suggestions?

### PR DESCRIPTION
**Pull request by dalyathan**
# Description

The changes in this PR implement and close #456 
# Breaking changes

The field `Product.customFields.primaryCollection` now has type `[String!]`
# Screenshots

CG
Fabrix

# Checklist

📌 Always:
- [X] I have set a clear title
- [X] My PR is small and contains a single feature
- [X] I have [checked my own PR](## "Fix typo's and remove unused or commented out code")

👍 Most of the time:
- [X] I have added or updated test cases for important functionality
- [X] I have updated the README if needed

📦 For publishable packages:
- [X] I have [increased the version number](## "After increasing the version to the next patch/minor/major, the package will be published automatically after merge") in `package.json`
- [X] I have added my changes to the `CHANGELOG.md` [See this example](https://github.com/Pinelab-studio/pinelab-vendure-plugins/blob/main/packages/vendure-plugin-invoices/CHANGELOG.md)

### CODE REVIEW

File: packages/vendure-plugin-primary-collection/src/api/primary-collection.resolver.ts [hunk 1]
```diff
@@ -1,13 +1,46 @@
...
+  @ResolveField('channels')
+  @Resolver('Collection')
+  async channels(
+    @Ctx() ctx: RequestContext,
+    @Parent() collection: Collection
+  ): Promise<Channel[] | undefined> {
+    await this.entityHydrator.hydrate(ctx, collection, {
+      relations: ['channels'],
+    });
+    return collection.channels;
```
[the remark this answers is not in the document]
Comment (dalyathan) [reply packages/vendure-plugin-primary-collection/src/api/primary-collection.resolver.ts:45]: why do we need to reverse the order of the decorators?

**Review by martijnvdbrug — CHANGES_REQUESTED** [review]
Impressive and clean solution :ok_hand: There is a bit of a performance hit when we load a big list of products + primary collections, because we do an extra DB call per primary collection. But I think there is no other way. And we will just test it out to see if it really matters.

I  do have some requested changes still:

* Can you make this a breaking change (major version)? A DB migration is needed, which looses all the currently set primary collections. 
* Is there any way we can migrate? For example, how would we install this plugin in the CG project, without loosing all the configured primary collections? It would be nice to have a script included that exports all the currently set primary collections to a JSON file, and then another script to insert them again.
* Can you also extend the graphql Product type with a field `breadcrumbs: CollectionBreadcrumb`? Same as a Collection. Because we have a primary collection, we can now resolve the breadcrumb of a product :smile: . It would prepend the id/name/slug of the product to the already existing breadcrumb of the primary collection.
* Would it make more sense to have the customField to be a `string` and put the entire JSON array in there? E.g. `[{"channelId":"1","collectionId":"5"}]`, instead of multiple strings of 1 JSON object. I did it this way (with `string[]`) in the CG configurator options as well, but I regret that decision... As a dev it's a bit easier to just be able to copy the entire custom field as JSON, instead of parsing each entry of an array. WDYT?
* Minor point: should we rename `getCandidatePrimaryCollections` to `getEligiblePrimaryCollections` to stay in line with Vendure's terminology?

File: packages/vendure-plugin-primary-collection/src/util/export-primary-collections.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      .getRawMany()) as ProductPrimaryCollectionPartial[];
+    const data: ExportedPrimaryCollectionData[] = [];
+    for (const product of allProducts) {
+      data.push({
+        productId: product.product_id,
+        primaryCollectionId: product.primaryCollection_id,
+      });
+    }
+    const jsonString: string = JSON.stringify(data, null, 2);
+    await fs.writeFile(EXPORTED_PRIMARY_COLLECTIONS_FILE, jsonString);
```
[the remark this answers is not in the document]
Comment (dalyathan) [reply packages/vendure-plugin-primary-collection/src/util/export-primary-collections.ts:37]: The directory will be the current directory, wherever it is

File: packages/vendure-plugin-primary-collection/src/api/primary-collection.resolver.ts [hunk 3]
```diff
@@ -1,13 +1,62 @@
...
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
+      (product.customFields as any).primaryCollection as string
+    );
+    if (collectionId) {
+      return this.collectionService.findOne(ctx, collectionId);
+    }
+  }
+
+  @ResolveField('breadcrumbs')
+  @Resolver('Product')
```
[the remark this answers is not in the document]
Comment (dalyathan) [reply packages/vendure-plugin-primary-collection/src/api/primary-collection.resolver.ts:38]: This won't work

File: packages/vendure-plugin-primary-collection/src/api/primary-collection.resolver.ts [hunk 4]
```diff
@@ -1,13 +1,62 @@
...
+@Resolver()
+export class PrimaryCollectionPluginResolver {
+  constructor(
+    private collectionService: CollectionService,
+    private entityHydrator: EntityHydrator
+  ) {}
+  @ResolveField('primaryCollection')
+  @Resolver('Product')
   async primaryCollection(
     @Ctx() ctx: RequestContext,
```
[the remark this answers is not in the document]
Comment (dalyathan) [reply packages/vendure-plugin-primary-collection/src/api/primary-collection.resolver.ts:24]: This won't work too

**Review by martijnvdbrug — APPROVED** [review]
Approved, nice work! I will create a task to test out the migration on the CG project. I've added 1 commit to also resolve the parent collection breadcrumb for the product:

https://github.com/Pinelab-studio/pinelab-vendure-plugins/pull/461/commits/b08cd0b02163c60f02eb473a041b8b422aac3a30

This should make it easier to create breadcrumbs on the storefront

image

**Pull request merged**